source/code/plugin/health/cluster_health_state.rb

```diff
@@ -0,0 +1,117 @@
+            @log.info "Making GET request to #{@uri.request_uri} @ #{Time.now.utc.iso8601}"
+            get_response = @http_client.request(get_request)
+            @log.info  "Got response of #{get_response.code} for #{@uri.request_uri} @ #{Time.now.utc.iso8601}"
+
+            if get_response.code.to_i == 404 # NOT found
+                #POST
+                update_request = Net::HTTP::Post.new(@uri.request_uri)
+                update_request["Content-Type"] = "application/json"
+
+            elsif get_response.code.to_i == 200 # Update == Patch
+                #PATCH
+                update_request = Net::HTTP::Patch.new(@uri.request_uri)
+                update_request["Content-Type"] = "application/merge-patch+json"
+            end
+            update_request["Authorization"] = "Bearer #{@token}"
+
+            update_request_body = get_update_request_body
+            update_request_body["state"] = state.to_json
+            update_request.body = update_request_body.to_json
+
+            @log.debug "Making an update request with #{update_request.method} "
+            update_response = @http_client.request(update_request)
+            @log.info "Got a response of #{update_response.code}"
+        end
+
+        def get_state
+            get_request = Net::HTTP::Get.new(@uri.request_uri)
+            get_request["Authorization"] = "Bearer #{@token}"
+            @log.info "Making GET request to #{@uri.request_uri} @ #{Time.now.utc.iso8601}"
+            get_response = @http_client.request(get_request)
+            @log.info  "Got response of #{get_response.code} for #{@uri.request_uri} @ #{Time.now.utc.iso8601}"
+
+            if get_response.code.to_i == 200
+                return JSON.parse(JSON.parse(get_response.body)["state"])
+            else
+                return {}
+            end
+        end
+
+        private
+        def get_token()
+            begin
+              if File.exist?(@token_file_path) && File.readable?(@token_file_path)
+                token_str = File.read(@token_file_path).strip
+                return token_str
+              else
+                @log.info ("Unable to read token string from #{@token_file_path}")
+                return nil
+              end
+            end
+        end
+
+        def get_http_client()
+            kube_api_server_url = get_kube_api_server_url
+            resource_uri = @@resource_uri_template % {
+                kube_api_server_url: kube_api_server_url
+            }
+            @uri = URI.parse(resource_uri)
+            http = Net::HTTP.new(@uri.host, @uri.port)
+            http.use_ssl = true
+            if !File.exist?(@cert_file_path)
+              raise "#{@cert_file_path} doesnt exist"
+            else
+              http.ca_file = @cert_file_path
+            end
+            http.verify_mode = OpenSSL::SSL::VERIFY_PEER
+            return http
+        end
+
+        def get_kube_api_server_url
```

Comment:
why is this function duplicated (from kubeapiclient) ?

Reply:
that constructs the entire URL based on resource which technically can construct this. but then it raises a hairy mess of dependency with oms_common, and I cannot unit test it locally.